### PR TITLE
[bitnami/grafana-operator] Custom update strategy for Grafana deployment

### PR DESCRIPTION
**Pull request by raviqqe**
**Description of the change**

This PR changes the` grafana-operator` Helm Chart so that it can update deployment even when persistence is enabled.

**Benefits**

We can get persistence working with the Helm Chart.

**Possible drawbacks**

We should revert this PR when the upstream repository of [Grafana Operator](https://github.com/grafana-operator/grafana-operator) is fixed.

**Applicable issues**

  - This is a workaround for https://github.com/grafana-operator/grafana-operator/issues/453.

**Additional information**

N/A

**Checklist** 

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])

### CODE REVIEW

**Comment by migruiz4** [comment]
Hi @raviqqe,

Thank you very much for your contribution!

I'd like to propose to you some changes to this PR.

Instead of adding a workaround intended to be reverted when the upstream is fixed, I'd appreciate it if you could add it as an additional feature `.Values.grafana.updateStrategy`, similar to `.Values.operator.updateStrategy`:

https://github.com/bitnami/charts/blob/f3329c07a6736ec0b6013f51b4d8c8bfad38024c/bitnami/grafana-operator/templates/deployment.yaml#L17-L19

That way the chart gains a new feature and it would be easier to maintain as we won't need to keep track of the upstream issue.

My recommendation would be to set `grafana.updateStrategy` empty by default and add a warning note next to `persistence.enabled` suggesting the change and referencing the upstream issue.

Let me know what you think.

**Comment by raviqqe** [comment]
@migruiz4 Sounds good! I'm gonna make the change.

**Comment by raviqqe** [comment]
BTW, why do we need to use the `Recreate` strategy for Grafana Operator containers? Volumes mounted for them are only the ones for service accounts and they can be shared among multiple containers as they are read only.

**Comment by raviqqe** [comment]
I'm still working on testing this PR. I'll let you know when it's ready for review.

**Review by migruiz4 — COMMENTED** [review]
Thank you very much for applying the all suggestions!

So far the PR looks good to me, let me know when you finish testing it.

I left some suggestions below, I would appreciate it if you could take a look at them

**Comment by migruiz4** [comment]
Hi @raviqqe,

Following up on this. 

Please let me know if you need any further help with this PR.

**Comment by raviqqe** [comment]
Sorry to be silent a while! I was busy with other part of my work. I'm gonna test the changes tomorrow and then this PR should be ready for review.

File: bitnami/grafana-operator/values.yaml [hunk 1]
```diff
@@ -552,6 +555,17 @@ grafana:
     timeoutSeconds: 5
     failureThreshold: 6
     successThreshold: 1
+  ## @param grafana.updateStrategy.type Set up update strategy for Grafana installation.
```
Comment (rafariossaa) [inline bitnami/grafana-operator/values.yaml:558]: There is an validation error on the  [metadata](https://github.com/bitnami-labs/readme-generator-for-helm#metadata) of the parameters in the values.yaml.

This should be:
```
  ## @param grafana.updateStrategy Set up update strategy for Grafana installation.
```

**Review by rafariossaa — CHANGES_REQUESTED** [review]

**Comment by rafariossaa** [comment]
LGTM

**Comment by StianHaugland1** [comment]
@raviqqe ive tried to set:
grafana:
  updateStrategy:
    type: Recreate
  
the grafana object get this:
deployment:
  strategy:
    type: Recreate
    
But the deployment still gets the old:
spec:
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
    
Am i doing something wrong?

**Comment by migruiz4** [comment]
Hi @StianHaugland1,

I'm sorry for the inconvenience, I think there has been some confusion on my side and we merged this feature although it hasn't been enabled in the upstream yet.

Here you can see how the latest stable version of grafana-operator (3.10.3) configures the deployment strategy:
https://github.com/grafana-operator/grafana-operator/blob/v3.10.3/pkg/controller/model/grafanaDeployment.go#L571-L605
```go
func getDeploymentSpec(cr *v1alpha1.Grafana, annotations map[string]string, configHash, plugins, dsHash string) v1.DeploymentSpec {
	return v1.DeploymentSpec{
		Replicas: getReplicas(cr),
		Selector: &v12.LabelSelector{
			MatchLabels: map[string]string{
				"app": GrafanaPodLabel,
			},
		},
		Template: v13.PodTemplateSpec{
			ObjectMeta: v12.ObjectMeta{
				Name:        GrafanaDeploymentName,
				Labels:      getPodLabels(cr),
				Annotations: getPodAnnotations(cr, annotations),
			},
			Spec: v13.PodSpec{
				NodeSelector:                  getNodeSelectors(cr),
				Tolerations:                   getTolerations(cr),
				Affinity:                      getAffinities(cr),
				SecurityContext:               getSecurityContext(cr),
				HostNetwork:                   getHostNetwork(cr),
				DNSPolicy:                     getDNSPolicy(cr),
				Volumes:                       getVolumes(cr),
				InitContainers:                getInitContainers(cr, plugins),
				Containers:                    getContainers(cr, configHash, dsHash),
				ServiceAccountName:            GrafanaServiceAccountName,
				TerminationGracePeriodSeconds: getTerminationGracePeriod(cr),
				PriorityClassName:             getPodPriorityClassName(cr),
			},
		},
		Strategy: v1.DeploymentStrategy{
			Type:          "RollingUpdate",
			RollingUpdate: getRollingUpdateStrategy(),
		},
	}
}
```

And this is how it is configured in the master branch, apparently to be released in the next major grafana-operator:4.0.0
https://github.com/grafana-operator/grafana-operator/blob/master/controllers/model/grafanaDeployment.go#L109-L118
```go
func getDeploymentStrategy(cr *v1alpha1.Grafana) v1.DeploymentStrategy {
	if cr.Spec.Deployment != nil && cr.Spec.Deployment.Strategy != nil {
		return *cr.Spec.Deployment.Strategy
	}


	return v1.DeploymentStrategy{
		Type:          "RollingUpdate",
		RollingUpdate: getRollingUpdateStrategy(),
	}
}
```

**Pull request merged**